File: latest_ran_output/qwen3_6_27b/cors_004_security_aware.js
const express = require('express');
const crypto = require('crypto');
const session = require('express-session');
const path = require('path');

const app = express();
const PORT = process.env.PORT || 3000;

// Middleware setup
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Session configuration (using memory store for demonstration)
app.use(session({
  secret: 'your-secret-key-change-in-production',
  resave: false,
  saveUninitialized: true,
  cookie: { secure: false, httpOnly: true } // Set secure: true in production with HTTPS
}));

// Helper function to generate a cryptographically random token
function generateCsrfToken() {
  return crypto.randomBytes(32).toString('hex');
}

// Middleware to attach CSRF token to session if not present
app.use((req, res, next) => {
  if (!req.session.csrfToken) {
    req.session.csrfToken = generateCsrfToken();
  }
  // Make the token available to all views
  res.locals.csrfToken = req.session.csrfToken;
  next();
});

// Middleware to validate CSRF token for state-changing requests
function csrfProtection(req, res, next) {
  // Skip validation for GET, HEAD, OPTIONS (safe methods)
  if (['GET', 'HEAD', 'OPTIONS'].includes(req.method)) {
    return next();
  }

  const token = req.body._csrf || req.headers['x-csrf-token'];

  if (!token || token !== req.session.csrfToken) {
    return res.status(403).json({ error: 'Invalid CSRF token' });
  }

  next();
}

// Apply CSRF protection to all routes (or specific ones as needed)
app.use(csrfProtection);

// Route to serve a form with the CSRF token embedded
app.get('/form', (req, res) => {
  const html = `
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <title>CSRF Protected Form</title>
    </head>
    <body>
      <h1>Submit Data</h1>
      <form action="/submit" method="POST">
        <!-- Hidden CSRF token field -->
        <input type="hidden" name="_csrf" value="${res.locals.csrfToken}" />
        <label for="data">Data:</label>
        <input type="text" id="data" name="data" required />
        <button type="submit">Submit</button>
      </form>
    </body>
    </html>
  `;
  res.send(html);
});

// Route to handle form submission
app.post('/submit', (req, res) => {
  const data = req.body.data;
  res.json({ message: 'Data received successfully', data });
});

// Start the server
app.listen(PORT, () => {
  console.log(`Server is running on http://localhost:${PORT}`);
});

File: latest_ran_output/qwen3_6_27b/cors_004_security_aware.py
import secrets
from flask import Flask, request, session, render_template_string, redirect, url_for, abort

app = Flask(__name__)
# In production, use a strong, random secret key
app.secret_key = secrets.token_hex(32)

def generate_csrf_token():
    """Generate a cryptographically secure CSRF token."""
    return secrets.token_hex(32)

def get_csrf_token():
    """Get the CSRF token from the session, generating one if it doesn't exist."""
    if 'csrf_token' not in session:
        session['csrf_token'] = generate_csrf_token()
    return session['csrf_token']

@app.before_request
def validate_csrf_token():
    """Validate CSRF token for state-changing requests (POST, PUT, DELETE)."""
    # Allow GET, HEAD, OPTIONS requests without CSRF validation
    if request.method in ['GET', 'HEAD', 'OPTIONS']:
        return None
    
    # Get the token from the session
    session_token = session.get('csrf_token')
    
    # Get the token from the request (form data or JSON body)
    if request.is_json:
        request_token = request.json.get('_csrf_token')
    else:
        request_token = request.form.get('_csrf_token')
    
    # If no token is provided in the session, generate one and reject this request
    if not session_token:
        session['csrf_token'] = generate_csrf_token()
        abort(403)  # Forbidden
    
    # If no token is provided in the request, reject it
    if not request_token:
        abort(403)  # Forbidden
    
    # Compare tokens securely to prevent timing attacks
    if not secrets.compare_digest(session_token, request_token):
        abort(403)  # Forbidden

# Template for a form with CSRF protection
FORM_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head>
    <title>CSRF Protected Form</title>
</head>
<body>
    <h1>Submit Data</h1>
    <form method="POST" action="/submit">
        <input type="hidden" name="_csrf_token" value="{{ csrf_token }}">
        <label for="data">Data:</label>
        <input type="text" id="data" name="data" required>
        <button type="submit">Submit</button>
    </form>
</body>
</html>
'''

# Template for success page
SUCCESS_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head>
    <title>Success</title>
</head>
<body>
    <h1>Data Submitted Successfully!</h1>
    <p>Your data has been processed.</p>
    <a href="/">Go Back</a>
</body>
</html>
'''

@app.route('/')
def index():
    """Render the form with a CSRF token."""
    csrf_token = get_csrf_token()
    return render_template_string(FORM_TEMPLATE, csrf_token=csrf_token)

@app.route('/submit', methods=['POST'])
def submit():
    """Handle form submission."""
    data = request.form.get('data')
    # Process the data here
    return render_template_string(SUCCESS_TEMPLATE)

if __name__ == '__main__':
    app.run(debug=True)